### PR TITLE
Emote overlays.

### PR DESCRIPTION
**Pull request by DominikPanic**
:cl: LaKiller8
imageadd: New overlay icon when throwing greandes created by Sleepy Retard.
tweak: Emoting now has a slightly different icon. Emote icons should now also appear above the HUD for clarity.
/:cl:

### CODE REVIEW

File: code/modules/mob/mob.dm [hunk 1]
```diff
@@ -794,8 +794,27 @@ mob/proc/yank_out_object()
...
+			continue
+		to_chat(M, emote_overlay)
+
+	if(remove_delay)
+		addtimer(CALLBACK(src, .proc/remove_emote_overlay, client, emote_overlay, viewers), remove_delay)
+
+
+/mob/proc/remove_emote_overlay(client/C, image/emote_overlay, list/viewers)
+	if(C)
+		C.images -= emote_overlay
```
Comment (Ghommie) [inline code/modules/mob/mob.dm]: `C?.images -= emote_overlay`, line above unneeded I guess.

File: code/modules/mob/mob.dm [hunk 2]
```diff
@@ -794,8 +794,27 @@ mob/proc/yank_out_object()
...
+	if(!viewers)
+		viewers = viewers()
+	for(var/mob/M in viewers)
+		if(!M.client)
+			continue
+		M.client.images -= emote_overlay
```
Comment (Ghommie) [inline code/modules/mob/mob.dm]: M.client?.images -= emote_overlay, two lines above unneeded.

File: code/modules/mob/mob.dm [hunk 3]
```diff
@@ -794,8 +794,27 @@ mob/proc/yank_out_object()
...
-/mob/proc/remove_emote_overlay(var/image/overlay_to_remove)
-	overlays -= overlay_to_remove
+/mob/proc/add_emote_overlay(image/emote_overlay, remove_delay = TYPING_INDICATOR_LIFETIME)
+	var/viewers = viewers()
+	for(var/mob/M in viewers)
+		if(M.stat == DEAD)
+			continue
+		to_chat(M, emote_overlay)
+
+	if(remove_delay)
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/modules/mob/mob.dm]: It's actually intended to be this way, the typing indicator has no delay and instead is removed whenever you close it again.

File: code/modules/mob/mob.dm [hunk 4]
```diff
@@ -794,8 +794,27 @@ mob/proc/yank_out_object()
 	return TRUE
 
 
-/mob/proc/remove_emote_overlay(var/image/overlay_to_remove)
-	overlays -= overlay_to_remove
+/mob/proc/add_emote_overlay(image/emote_overlay, remove_delay = TYPING_INDICATOR_LIFETIME)
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/modules/mob/mob.dm]: This is for audible emotes only.

[the remark this answers is not in the document]
Comment (Ghommie) [reply code/modules/mob/mob.dm]: Yea, but if deafs can't hear the audible message, they shouldn't see the overlay either, no?

File: code/modules/mob/mob.dm [hunk 5]
```diff
@@ -794,8 +794,27 @@ mob/proc/yank_out_object()
...
+	var/viewers = viewers()
+	for(var/mob/M in viewers)
+		if(M.stat == DEAD)
+			continue
+		to_chat(M, emote_overlay)
```
[the remark this answers is not in the document]
Comment (DominikPanic) [reply code/modules/mob/mob.dm]: Not sure tbh I copied this snippet over from the emote code that seems to work well.

File: code/modules/mob/mob.dm [hunk 6]
```diff
@@ -794,8 +794,24 @@ mob/proc/yank_out_object()
...
+		if(!isobserver(M) && (M.stat != CONSCIOUS || isdeaf(M)))
+			continue
+		to_chat(M, emote_overlay)
+
+	if(remove_delay)
+		addtimer(CALLBACK(src, .proc/remove_emote_overlay, client, emote_overlay, viewers), remove_delay)
+
+
+/mob/proc/remove_emote_overlay(client/C, image/emote_overlay, list/viewers)
+	C?.images -= emote_overlay
```
Comment (Rohesie) [inline code/modules/mob/mob.dm]: This will require to be inside an `if(C)`

File: code/modules/mob/mob.dm [hunk 7]
```diff
@@ -794,8 +794,24 @@ mob/proc/yank_out_object()
...
+	if(!viewers)
+		viewers = viewers()
+	for(var/mob/M in viewers)
+		M.client?.images -= emote_overlay
```
Comment (Rohesie) [inline code/modules/mob/mob.dm]: Same here for `M.client`

File: code/modules/mob/mob.dm [hunk 8]
```diff
@@ -794,8 +794,24 @@ mob/proc/yank_out_object()
...
+			continue
+		to_chat(M, emote_overlay)
+
+	if(remove_delay)
+		addtimer(CALLBACK(src, .proc/remove_emote_overlay, client, emote_overlay, viewers), remove_delay)
+
+
+/mob/proc/remove_emote_overlay(client/C, image/emote_overlay, list/viewers)
+	C?.images -= emote_overlay
+	if(!viewers)
```
Comment (Rohesie) [inline code/modules/mob/mob.dm]: This should probably not be a thing. If no viewers are selected, it probably shouldn't be looking for more. This should be defined at the call level.

File: code/modules/mob/mob.dm (hunk 6 above)
Comment (DominikPanic) [reply code/modules/mob/mob.dm]: Thanks Ghom

File: code/modules/mob/mob.dm (hunk 7 above)
Comment (DominikPanic) [reply code/modules/mob/mob.dm]: Thanks Ghom

File: code/modules/mob/mob.dm (hunk 1 above)
Comment (Rohesie) [reply code/modules/mob/mob.dm]: That'd be bad. If `C` doesn't exist that will create an operation that looks like "`null` equals something".

File: code/modules/mob/mob.dm (hunk 2 above)
Comment (Rohesie) [reply code/modules/mob/mob.dm]: Same here.

File: code/modules/mob/mob.dm (hunk 1 above)
Comment (Ghommie) [reply code/modules/mob/mob.dm]: Oh, right there, I forgot we had this kind of talk long time ago.

**Pull request merged**